### DISCUSSION TITLE
Example of using Pandas to specify stress period data

### DISCUSSION THREAD

**Discussion by mbakker7** [modflowpy/flopy#2135] [General]
If I understand correctly, you can specify stress period data (e.g. for wells, ghbs, chds, etc) now with a Pandas DataFrame. Is there an example (notebook) where this is done? I could not find one. Or is there maybe detailed documentation on how to do this?
Thanks in advance.

**Comment by wpbonelli** [comment]
Initial support for dataframes came in with https://github.com/modflowpy/flopy/commit/be104c4a38d2b1f4acfbdd5a49a5b3ddb325f852. Besides docstrings we don't have specific docs on it, but the tests updated in that commit show a few samples. After taking a second look I found a few holes in the coverage, to be fixed by #2136. I also updated `mf6_mnw2_tutorial01.py` to provide stress period data as a dataframe.

**Reply by mbakker7** [reply]
Thanks, @wpbonelli. That helps. Couple of follow up questions. From your post I understand that for a GHB cell, I now have to do:
```
   riv_spd = pd.DataFrame(
       [[0, 0, 0, 0, 101.0, 10.0, 100.0], 
       [0, 0, 0, 1, 80.0, 10.0, 90.0]],
       columns=["per", "k", "i", "j", "stage", "cond", "rbot"],
       )

   pers = riv_spd.groupby("per")
   riv_spd = {i: pers.get_group(i).drop("per", axis=1) for i in [0]}
```
Questions:
1. I presume the columns must have the prescribed names `"k", "i", "j", "stage", "cond"`. Is there a way in flopy to get an empty DataFrame which already has columns with those names (and possibly the data type for those columns)? That would be very helpful.
2. When specifying the stress period data, I still have to convert the DataFrame into a dictionary where the keys are the numbers of the stress period. But the DataFrame that is created above it already has a column named `per`. Any reason why Flopy cannot accept a DataFrame as the stress period data where it needs a column named `per` for the number of the stress period? Then Flopy can do the two last lines of the code above behind the scenes. No reason to trouble the user with having to do that over and over.

**Reply by wpbonelli** [reply]
> I presume the columns must have the prescribed names "k", "i", "j", "stage", "cond". Is there a way in flopy to get an empty DataFrame which already has columns with those names (and possibly the data type for those columns)? That would be very helpful.

The package classes have static methods providing the dtype, so one way (in context above) is

```python
pd.DataFrame(np.empty(0, dtype=flopy.modflow.ModflowRiv.get_default_dtype()))
```

I could see the case for an e.g. `empty()` convenience method on the packages. 

> When specifying the stress period data, I still have to convert the DataFrame into a dictionary where the keys are the numbers of the stress period. But the DataFrame that is created above it already has a column named per. Any reason why Flopy cannot accept a DataFrame as the stress period data where it needs a column named per for the number of the stress period? Then Flopy can do the two last lines of the code above behind the scenes. No reason to trouble the user with having to do that over and over.

I think it would make sense to support a `per` (or maybe `kper`) column in recarray or dataframe input, as an alternative to a dict. I think it will require changes in `flopy.utils.MfList` [here](https://github.com/modflowpy/flopy/blob/4c717cbfe189c7fd74ba85d0839e4df575b56d80/flopy/utils/util_list.py#L301)

**Reply by christianlangevin** [reply]
I like the idea of a kper column in the dataframe/recarray.  We have been kicking that idea around for a while now and should see what it would take to support.

**Reply by wpbonelli** [reply]
#2179 adds support for a `kper` column
